Added `math product` support

gdhuper opened the pull request:
## Description

Added `math product` to multiply numbers together 

eg: 
```bash
> echo [1 2 3] | math product
> 6
```
This PR will close issue : #2081 

## Checklist
- [x] Code compiles 
- [x] Fix any lint/clippy issues
- [x] Fix any merge conflicts
- [x] Add unit tests / e2e tests
- [x] Add/update documentation

gdhuper commented:
Hi @jonathandturner, could you take a look and provide feedback on this draft? Thanks

jonathandturner commented:
@gdhuper - yup looks good so far

gdhuper commented:
I noticed something odd when testing my local branch for math operations.
I can run all math operations fine on my local build if use `math <operation>` but it won't work on my globally installed nushell. I tried updating it, but it still won't work. Is there something I am missing?
```bash
❯ echo [1 2 3 4 5] | math sum
error: Command not found
- shell:1:27
1 | echo [1 2 3 4 5] | math sum
  |                    ^^^^ command not found
nushell on 📙 gdhuper/math_product is 📦 v0.17.0 via 🦀 v1.43.1
❯ echo [1 2 3 4 5] | sum
15
nushell on 📙 gdhuper/math_product is 📦 v0.17.0 via 🦀 v1.43.1
❯ echo [1 2 3 4 5] | avg
error: Command not found
- shell:1:22
1 | echo [1 2 3 4 5] | avg
  |                    ^^^ command not found
nushell on 📙 gdhuper/math_product is 📦 v0.17.0 via 🦀 v1.43.1
❯ echo [1 2 3 4 5] | math avg
error: Command not found
```

jonathandturner commented:
@gdhuper - can you run `version` in your globally installed nushell and see what the version is?

gdhuper commented:
> @gdhuper - can you run `version` in your globally installed nushell and see what the version is?
I was running nushell version 0.13.0
`cargo update -p nu` didn't update it to the latest version, I had to do this `cargo install nu -version 0.17.0` to update the version.

jonathandturner commented:
@gdhuper - Yeah, I think `cargo install` does the installing, while `cargo update` updates the dependencies of the current project. I agree those might be a confusing combination.

Your checklist looks great. Let us know if you have any questions

gdhuper commented:
@jonathandturner I am running into this issue when I attempt to calculate `product` of `FileSize` primitive (`ls | select size | math product`) and it throws:

`thread 'main' panicked at 'attempt to multiply with overflow'`.

If I am not mistaken, I think the product exceeds max value for `u64` and causes overflow. Here is the [code](https://github.com/gdhuper/nushell/blob/acb9d3128a79ebb260f2736c4a4a84ced9484f08/crates/nu-protocol/src/value/primitive.rs#L223) that I added for supporting `Mul` for `FileSize` that is causing the error. Do you have any suggestions on how to solve this? Thanks

thegedge commented:
> Do you have any suggestions on how to solve this? Thanks

I think we have two options:

1. Switch from `u64` to `BigInt`.
2. Use https://doc.rust-lang.org/std/primitive.u64.html#method.checked_mul and produce an error if `None` is returned (in other words, if there's an overflow).

(1) is a pretty substantial change that may have implications we're not aware of, so I'm inclined to do (2) for now until we can can come to a consensus on whether or not we want to do (1). We can always come back later and change this 🙂

gdhuper commented:
> > Do you have any suggestions on how to solve this? Thanks
> 
> I think we have two options:
> 
> 1. Switch from `u64` to `BigInt`.
> 2. Use https://doc.rust-lang.org/std/primitive.u64.html#method.checked_mul and produce an error if `None` is returned (in other words, if there's an overflow).
> 
> (1) is a pretty substantial change that may have implications we're not aware of, so I'm inclined to do (2) for now until we can can come to a consensus on whether or not we want to do (1). We can always come back later and change this 🙂

Thanks @thegedge for the suggestions. Yes, you are right, changing `u64` to `BigInt` might have other implications. I'll try out 2) and see I am able to get around the problem.

jonathandturner commented:
@gdhuper - how's the PR going? Feel free to reach out if you need anything.

gdhuper commented:
> @gdhuper - how's the PR going? Feel free to reach out if you need anything.

Sorry about inactivity on this PR. Was swamped with work stuff last week. Planning on wrapping up this PR end of this week. Thank you for reach out

gdhuper commented:
Apologies for inactivity on this PR again. I have rebased this branch off of main branch and refactored my code accordingly. Any feedback will be appreciated.

gdhuper commented on crates/nu-cli/src/commands/math/product.rs:
```diff
@@ -0,0 +1,128 @@
+            )
+        }),
+
+        v if v.is_none() => prod(
```
Yes, so I pretty much followed `math sum` implementation and didn't think of the implications of this (given my naivety with the codebase). Since I first did pass on `math product`, looks like a lot of the code has been refactored and I was just trying to achieve parity with `math sum`. 

As you mentioned, would it make more sense to throw error instead of interpreting `Primitive::Nothing`?

thegedge replied:
> As you mentioned, would it make more sense to throw error instead of interpreting `Primitive::Nothing`?

Yeah, I think we want to lean towards this approach. Again, don't worry about it in this PR 🙂 

The reason for it is that we could silently ignore a `Nothing` in a stream that the end user doesn't expect. In that scenario, our math functions would happily still produce a result, even though it's wrong.

I'd love to make it so that we can be explicit in that kind of way, but still allow our users to override that behaviour. For example (this won't work today):

```
alias math sum [] { default column 0 | math sum }
```

gdhuper commented on crates/nu-cli/src/commands/math/product.rs:
```diff
@@ -0,0 +1,128 @@
+            UntaggedValue::int(1).into_untagged_value(),
+            values
+                .to_vec()
+                .iter()
```
Oh gotcha. let me try to see if I can get it to work without `to_vec`.

gdhuper commented on crates/nu-cli/src/commands/math/product.rs:
```diff
@@ -0,0 +1,128 @@
+        .ok_or_else(|| {
+            ShellError::labeled_error(
+                "could not convert to big decimal",
+                "could not convert to big decimal",
```
Will remove. Thanks

jonathandturner commented:
Looks good!
